[DO NOT MERGE] Enable the new landing screen in WP app and fix UI tests

ovitrif opened the pull request:
Fixes n/a

> **Warning** Don't merge this PR please... _at least not yet!_.

I created this PR to run the UI tests as a starting point to investigate how to resolve the failures caused by the `Compose` implementation of the new landing screen.


To test:
- **Verify** the `*.instrumented-tests` step in CI is green 🟢 

## Regression Notes
1. Potential unintended areas of impact
   Any UI tests going through the login flow, in both apps: WP & JP.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   CI.

3. What automated tests I added (or what prevented me from doing so)
   N/a

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Enable the new landing screen in WP app

ovitrif committed:
Add dependencies for testing Jetpack Compose UI

ovitrif committed:
Add utility to help writing compose UI tests in Java

ovitrif committed:
Add documentation for compose utility function

ovitrif committed:
Deps: Move compose ui testing above tooling

ovitrif committed:
Adapt login e2e tests to the compose ui in revamped landing screen

ovitrif committed:
Move ComposeTestRule to BaseTest

ovitrif committed:
Add super to mComposeTestRule references for readability

ovitrif committed:
Adapt BaseTest.wpLogin to the compose ui in revamped landing screen

This change affects UI tests in:
- BlockEditorTests
- ReaderTests
- StatsTests
- JPScreenshotTest
- WPScreenshotTest

ovitrif committed:
Adapt ContactUs e2e tests to the compose ui in new landing screen

ovitrif committed:
Adapt signup e2e tests to the compose ui in revamped landing scree

mkevins reviewed (approved):
Thanks for tackling the changes needed for fixing the UI tests for Compose :+1: 

I tested this, and it is working nicely. Also, the code changes make sense to me. I also ran the UI tests manually and observed that the automation was working correctly.

jostnes commented:
Thanks for updating the tests @ovitrif! I was pinged as part of the `mobile-ui-testing-squad` group, I tested it out locally and it works well. Coincidentally, I am also working on updating the Jetpack UI tests (Ref: pdcxQM-1Nc-p2 ) which also has a compose screen so this is really helpful! 

I was chatting with @ParaskP7 about this recently and one of the suggestions is to start by updating the java test files to kotlin. In this PR adding a new `.../ComposeUiTestingUtils.kt` seems to work to get access to the Compose test APIs and to perform the click needed to interact with the button on the test. @ParaskP7 what do you think about using this workaround and updating the java test files to kotlin later (which I agree still needs to be done)?

ovitrif commented:
> Thanks for updating the tests @ovitrif! I was pinged as part of the `mobile-ui-testing-squad` group, I tested it out locally and it works well. Coincidentally, I am also working on updating the Jetpack UI tests (Ref: pdcxQM-1Nc-p2 ) which also has a compose screen so this is really helpful!

Thank you @jostnes for validating the changes and reporting it worked for you locally 🙇 !
I didn't know about the project you referenced, nice share 🙏 , I added an internal comment (ref pdcxQM-1Nc-p2#comment-1570).

> I was chatting with @ParaskP7 about this recently and one of the suggestions is to start by updating the java test files to kotlin. In this PR adding a new `.../ComposeUiTestingUtils.kt` seems to work to get access to the Compose test APIs and to perform the click needed to interact with the button on the test. @ParaskP7 what do you think about using this workaround and updating the java test files to kotlin later (which I agree still needs to be done)?

That's nice 😉 , the reason I added `ComposeUiTestingUtils.kt` was exactly for that: to delay having to convert the Java test files to Kotlin 😅 so that I can keep the diff of these changes focused only on actually fixing the tests. Given that this PR should only be merged after the `21.7` release branch is cut, with this approach we can still cherry-pick the needed commits that add support for testing Compose UI 👍

ParaskP7 commented:
Great work @ovitrif ! ❤️ 💯 🥇 

> I was chatting with @ParaskP7 about this recently and one of the suggestions is to start by updating the java test files to kotlin. In this PR adding a new .../ComposeUiTestingUtils.kt seems to work to get access to the Compose test APIs and to perform the click needed to interact with the button on the test.

True, nice timing! 👍 

> @ParaskP7 what do you think about using this workaround and updating the java test files to kotlin later (which I agree still needs to be done)?

@jostnes this is a nice workaround indeed and if you feel comfortable with it I too recommend delaying the Java-to-Kotlin conversion for later. 👍 

However, I wouldn't want us to forget about it and would still strive to do this conversion sooner than later, that is, given the fact that:
- This UI tests conversion should be straightforward, plus, almost risk-free, and 🎯
- It will remove the need for this extra layer of Compose utilities that now engineers need to be aware of when working with Compose related UI tests. 🤔

> That's nice 😉 , the reason I added ComposeUiTestingUtils.kt was exactly for that: to delay having to convert the Java test files to Kotlin 😅 so that I can keep the diff of these changes focused only on actually fixing the tests.

👍 

PS: I am also glad that our current Compose `1.1.1` version works for our current Compose UI tests needs, which is another win, since unfortunately we need to wait for `compileSdkVersion = 33` upgrade before updating Compose to any latest version, even the `1.2.X`... 🤷

> Given that this PR should only be merged after the 21.7 release branch is cut, with this approach we can still cherry-pick the needed commits that add support for testing Compose UI 👍

🎉

ovitrif commented:
Thank you @ParaskP7 for your input 💯 🏅 

> However, I wouldn't want us to forget about it and would still strive to do this conversion sooner than later, that is, given the fact that:
> 
> * This UI tests conversion should be straightforward, plus, almost risk-free, and 🎯
> * It will remove the need for this extra layer of Compose utilities that now engineers need to be aware of when working with Compose related UI tests. 🤔

I can help with this too 👍 ; and I agree we should have it on our plans sooner rather than later. There's a lot to gain in terms of cleaner test code, one good example is the last point you already mentioned 🙇 

I wanted a confirmation from a discussion on this topic before going all in with Java to Kotlin conversion for the UI tests code 😅 . Now it happened 🚀

ParaskP7 commented:
> I wanted a confirmation from a discussion on this topic before going all in with Java to Kotlin conversion for the UI tests code 😅 . Now it happened 🚀

❤️ x ❤️ x ❤️ = 🚀

PS: @ovitrif I am not sure what kind of confirmation you wanted, on this or any topic just like that, but as far I am concerned please go all-in with every such tech-debt related topic, without any confirmation at all, I am 100% with you and will back you up all the way no matter, you all are the holders and owners of your own tech-debt, so please, go for it heads-down any and every time you need/have/want to! 💯

ovitrif commented:
> > I wanted a confirmation from a discussion on this topic before going all in with Java to Kotlin conversion for the UI tests code 😅 . Now it happened 🚀
> 
> ❤️ x ❤️ x ❤️ = 🚀
> 
> PS: @ovitrif I am not sure what kind of confirmation you wanted, on this or any topic just like that, but as far I am concerned please go all-in with every such tech-debt related topic, without any confirmation at all, I am 100% with you and will back you up all the way no matter, you all are the holders and owners of your own tech-debt, so please, go for it heads-down any and every time you need/have/want to! 💯

Thank you for your support @ParaskP7 😍 , you are awesome 🏅 🏆 🚀 

P.S. I wanted to make sure contributors are ok with converting Java UI tests to Kotlin. I didn't see a lot of Kotlin in the UI testing code and thought perhaps there might have been a preference for Java 😅

ovitrif commented:
Closed in favor of:
- https://github.com/wordpress-mobile/WordPress-Android/pull/17869
- https://github.com/wordpress-mobile/WordPress-Android/pull/17871
